keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,283 @@
+- group and version
+- subresource
+- output media type
+However, group, version and subresource are reflected in the `SelfLink` of
+returned object. As a result, for a given (potentially unversioned) object,
+we can identify all its possible serializations by (SelfLink, media-type)
+pairs, and that is what we will do below.
+
+We propose introducing an internal type implementing `runtime.Object` interface,
+that is encapsulating the original object and is additionally able to accumulate
+its serialized versions. It will look like this:
+```
+// TODO: Better name is welcome :)
+type CachingObject struct {
+	// Object is the object (potentially in the internal version)
+	// for which serializations should be cached for future reuse.
+	Object Object
+	// Serializations is a list of cached serializations of Object.
+	//
+	// TODO: Maybe it should be a map?
+	Serializations []*Serialization
+}
+
+// TODO: Better name is welcome :)
+type Serialization struct {
+	// selfLink of the serialized object identifying endpoint
+	// (e.g. group, version, subresource)
+	selfLink string
+	// serializations is a list of cached serializations for different
+	// media types.
+	//
+	// TODO: Maybe explicitly: jsonSerialization and protobufSerialization?
+	serializations []*serialization
+}
+
+// TODO: Better name is welcome :)
+type serialization struct {
+	once sync.Once
+	// mediaType to which the object is serialized.
+	mediaType string
+	// raw is serialized object.
+	raw []byte
+	// err is error from serialization.
+	err error
+
+	// object is a temporary field that is set only during serialization to
+	// avoid changing any existing interfaces.
+	//
+	// TODO: This probably should be part of `Serialization` instead.
+	object runtime.Object
+}
+```
+
+With those objects, we need to then slightly modigy the generic apimachinery
+infrastructure for serializers:
+- `versioning.codec` when serializing `CachingObject` chooses appropriate
+`Serialization` object from its `Serializations` field and passes it down;
+if such `Serialization` doesn't exist, it simply deep-copies the wrapped
+`Object` field
+- json and protobuf serializers, serialize and cache it in the object only
+on the first serialization (`once` field) and return the cached result
+
+Now the question is what is responsible for creating `CachingObject`.
+In the first attempty, we will put that responsibility into `watchCache`:
+when receiving an event via watch from etcd it will be opaquing it into
+`CachingObject` and operating on objects of that type later.
+
+Note that given that objects are already opaqued in `watchCache`, we
+also avoid unnecessary serialization when GET/LIST requests are served
+from cache.
+
+One open questions is how do we decide what exactly we want to cache.
+Caching everything stored in `watchCache` seems expensive from memory
+usage point of view. What we suggest, is allow configuring individual
+`registries` with exact versions that we want to cache and propagate
```

Comment:
Another idea: each caching object, on construction, is registered with a per-resource LRU. The LRU deletes cached serializations when it evicts objects. Objects also remove themselves when they are dropped from the watch cache. Something like this.

Reply:
So let me clarify one thing - the "registering" part is used to ensure that we will really serialize object once, right?
The non-obvious part here is that, given our identifier is selfLink (or at cache level, probably (selfLink, RV) pair), we don't know the selfLink at the point of creation. So it seems that we would conceptually register a tree with root being RV, to which we can add selfLink-identified children, and media-type below those.

But at the high-level, I think it's basically variation about LRU option, which I think is a useful variation, though I'm not 100% sure it will actually help that much with complexity of the solution. I will think about it a bit more.